### ISSUE TITLE
Ant Media Server Enterprise 2.2.1 crash when call REST API /v2/broadcasts/{id}/rtmp-endpoint

### ISSUE THREAD

**Issue by ipliu** [ant-media/Ant-Media-Server#3201]
### Short description
I publish a WebRTC live stream to AMS and want to publish this stream to another server via RTMP.
I use REST API /v2/broadcasts/{id}/rtmp-endpoint to do this.
When published WebRTC live stream, call REST API /v2/broadcasts/{id}/rtmp-endpoint to publish to another server.
But if I stop publish the WebRTC live stream immediately after the stream published, the AMS crashed.

### Environment

* Operating system and version: Ubuntu 18.04.5 LTS
* Java version: 11.0.10
* Ant Media Server version: v2.2.1
* Browser name and version: Chrome 90

### Steps to reproduce
1. Modify webapps/LiveApp/index.html, when receive publish_started callback, call REST API /v2/broadcasts/{id}/rtmp-endpoint
2. Use a browser to open {server_url}/LiveApp/index.html
3. Click Start Publishing button, when Stop Publishing button is enable, click it immediately.
4. Repeat step 3 and AMS will crash.

### Expected behavior
AMS publish RTMP stream to another server.

### Actual behavior
AMS crashed

 

### Logs
ant-media-server.log  https://pastebin.com/Kxq1zb5q
hs_err_pid11859.log  https://pastebin.com/h6FcjkYu

**Comment by mekya** [comment]
Thank you. We have fixed a similar case in 2.3.1 or 2.3.2 but let us check your case as well.

This issue will be scheduled in this week's sprint

**Comment by mekya** [comment]
Sorry I could not schedule this issue in this sprint. Because previous issues are not finished and there were no much space for the new issues.

**Comment by cemklc** [comment]
Hello,

I tried to reproduce this scenario on AMS v2.3.2 using 2 different Ant Media servers. I published a webrtc stream from server1 LiveApp/index.html and added an rtmp-endpoint which is created on server2. I tried clicking stop publishing as soon as it is available however I didn't encounter any crash. Also the stream was clear when viewed on server2.

As @mekya stated, it might be solved for the latest releases. We highly recommend updating AMS version if it is convenient for you, since there are also another memory and cpu optimizations were added on v2.3 and many bugs are fixed for v2.3.1 and v2.3.2.

**Comment by ipliu** [comment]
Hi,

This issues reproducible steps are as follow (on AMS v2.3.2).
1. Make sure that the client can call the server1 LiveApp REST API.
2. Add XMLHttpRequest to call REST API to add rtmp-endpoint when received callback 'publish_started' in server1 LiveApp/index.html.
3. Publish a webrtc stream from server1 LiveApp/index.html and clicking stop publishing as soon as it is available.
4. Repeat step 3 several times, server1 AMS will crash. If server1 AMS start as service, it will auto restart.
5. I tried AMS v2.3.2 on Ubuntu 18.04.5 and CentOS 8.3.2011 and this issue is occurred on both OS.

And here is my XMLHttpRequest code in LiveApp/index.html.
```
var httpRequest = new XMLHttpRequest();
httpRequest.onreadystatechange = function(){
	if (httpRequest.readyState === XMLHttpRequest.DONE) {
		if (httpRequest.status === 200) {
			console.log(httpRequest.responseText);
		} else {
			console.log('Error:' + httpRequest.status);
		}
	}
};
httpRequest.open('POST', appName + 'rest/v2/broadcasts/' + obj.streamId + '/rtmp-endpoint');
httpRequest.setRequestHeader('Content-Type', 'application/json');
httpRequest.send('{"rtmpUrl":"rtmp://server2' + appName + obj.streamId + '"}');
```

**Comment by ipliu** [comment]
Here is my logs.

[ant-media-server_pid1415276.log](https://github.com/ant-media/Ant-Media-Server/files/6583539/ant-media-server_pid1415276.log)
[hs_err_pid1415276.log](https://github.com/ant-media/Ant-Media-Server/files/6583541/hs_err_pid1415276.log)

**Comment by cemklc** [comment]
Thanks @ipliu for sharing the details. I have confirmed the crash and also we have found the issue which is related with a race condition between two threads. I tested for various cases and if I see the lines that I give below which states that the endpoint status is updated as broadcasting, then there won't be any crashes.

[vert.x-worker-thread-14] io.antmedia.muxer.MuxAdaptor - Endpoint status updated to broadcasting  for streamId: stream1 for url: rtmp://server2/LiveApp/stream1


For a quick fix, I have added a simple sleep function after http request in index.html file

""""""""""""""""
httpRequest.open('POST', appName + 'rest/v2/broadcasts/' + obj.streamId + '/rtmp-endpoint');
httpRequest.setRequestHeader('Content-Type', 'application/json');
httpRequest.send('{"rtmpUrl":"rtmp://server2' + appName + obj.streamId + '"}');
pauseBrowser(3000);
"""""""""""""""""
It makes the browser wait for 3 seconds so that the request is send and processed properly. Stopping the stream during somewhere in between process causes the crash. We are testing the issue on backend and it seems to be working properly now and most likely will be included in the next release. 

For now, if it is convenient for you, you can use the quickfix on index.html. I have used 3000 ms (note that the input is in milliseconds) on my end after testing several times, however you might need to change it to a more suitable value depending on your server. 

You can find the pauseBrowser function below.



--------------
	function pauseBrowser(millis) {
		var date = Date.now();
		var curDate = null;
		do {
			curDate = Date.now();
		} while (curDate-date < millis);
	}
------------

**Comment by mekya** [comment]
The related PR is merged into the release branch and master branch. It'll be available in 2.3.3

**Comment by ipliu** [comment]
Is this issue fixed in 2.3.3?

I tried AMS v2.3.3 on CentOS 8.3.2011 and this issue is occurred too.

Here is my server logs.

[ant-media-server_pid1529022.log](https://github.com/ant-media/Ant-Media-Server/files/6621171/ant-media-server_pid1529022.log)
[hs_err_pid1529022.log](https://github.com/ant-media/Ant-Media-Server/files/6621172/hs_err_pid1529022.log)

**Comment by mekya** [comment]
We thought we fixed that but we are wrong. I'm moving the issue to the Sprint again

**Comment by golgetahir** [comment]
Hi @ipliu ,

The issue is caused by a thread race, when the RTMP thread initializes outputFormatContext, then it continues to try writing the header and it takes a while. When you stop the broadcast in that interval, webRTC thread jumps in and calls writeTrailer method for all of the muxers. Since you stop the broadcast before RTMP thread finishes writeHeader, then crash happens when webRTC thread calls writeTrailer because there are no headers. A simple fix is just checking if the header is written like the above PR. Thank you for your attention to the details and testing our product with different test cases, this one is slipped away from us, it is fixed now. I hope I clearly explained the situation for you.

Cheers.

**Comment by ipliu** [comment]
Is this issue fixed in 2.3.3.1?

**Comment by mekya** [comment]
Hi @ipliu ,

No it's not fixed in 2.3.3.1.  It's fixed in the snapshots. You can get the latest snapshot and try that. 

https://antmedia.io/latest-snapshot/
